sg/SG_Node.3
.\"
.\" Copyright (c) 2010-2022 Julien Nadeau Carriere <vedge@csoft.net>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 21, 2022
.Dt SG_NODE 3
.Os Agar 1.7
.Sh NAME
.Nm SG_Node
.Nd Agar-SG base node class
.Sh SYNOPSIS
.Bd -literal
#include <agar/core.h>
#include <agar/sg.h>
.Ed
.Sh DESCRIPTION
.Nm
is the base class for any element of a
.Xr SG 3
scene.
This includes geometrical elements (e.g.,
.Xr SG_Object 3 )
as well as non-geometrical elements (e.g.,
.Xr SG_Light 3 ,
.Xr SG_Camera 3 ) .
.Pp
Nodes are organized in a tree structure, and the position of a
.Nm
in relation to its parent is determined by a 4x4 transformation matrix
(see
.Xr M_Matrix 3 ) .
.Sh INHERITANCE HIERARCHY
.Xr AG_Object 3 ->
.Nm .
.Sh INTERFACE
.nr nS 1
.Ft "void"
.Fn SG_RegisterClass "SG_NodeOps *ops"
.Pp
.Ft "void"
.Fn SG_NodeInit "void *node" "const char *name" "const SG_NodeOps *ops" "Uint flags"
.Pp
.Ft "void *"
.Fn SG_NodeAdd "void *pnode" "const char *name" "const SG_NodeOps *ops" "Uint flags"
.Pp
.Ft "void"
.Fn SG_NodeAttach "void *pnode" "void *node"
.Pp
.Ft "void"
.Fn SG_NodeDetach "void *pnode" "void *node"
.Pp
.Ft "void"
.Fn SG_NodeDraw "SG *sg" "SG_Node *node" "SG_View *view"
.Pp
.nr nS 0
The
.Fn SG_RegisterClass
function registers a new node class, described by the given
.Ft SG_NodeOps
structure.
.Pp
The
.Fn SG_NodeInit
function initializes the given
.Ft SG_Node
structure.
It is usually invoked from node constructor functions.
The
.Fa name
argument is a string identifier for the node.
.Fa ops
points to the
.Ft SG_NodeOps
structure which contains class information.
The
.Fa flags
argument should be 0.
The
.Fn SG_NodeAdd
variant also allocates, initializes and attaches the node to a parent node.
.Pp
The
.Fn SG_NodeAttach
and
.Fn SG_NodeDetach
functions attach/detach a node to/from a given parent.
.Pp
The
.Fn SG_NodeDraw
function is used to render a node to the display.
It is normally only invoked from the draw operation of
.Xr SG_View 3
(or another visualization widget derived from
.Ft SG_View ) .
.Fn SG_NodeDraw
assumes that the node's transformation matrix has already been applied
to the current viewing matrix.
.Sh NODE TRANSFORMATIONS
The following calls multiply a node's transformation matrix
.Va T
with a translation, scaling or rotation matrix.
Note that the
.Va T
matrix may also be manipulated directly with the
.Xr M_Matrix 3
interface.
.Pp
.nr nS 1
.Ft "void"
.Fn SG_Identity "SG_Node *node"
.Pp
.Ft "void"
.Fn SG_Translate "SG_Node *node" "M_Real x" "M_Real y" "M_Real z"
.Pp
.Ft "void"
.Fn SG_Translatev "SG_Node *node" "M_Vector3 v"
.Pp
.Ft "void"
.Fn SG_TranslateX "SG_Node *node" "M_Real x"
.Pp
.Ft "void"
.Fn SG_TranslateY "SG_Node *node" "M_Real y"
.Pp
.Ft "void"
.Fn SG_TranslateZ "SG_Node *node" "M_Real z"
.Pp
.Ft "void"
.Fn SG_Scale "SG_Node *node" "M_Real s"
.Pp
.Ft "void"
.Fn SG_Rotatev "SG_Node *node" "M_Real theta" "M_Vector3 axis"
.Pp
.Ft "void"
.Fn SG_RotateI "SG_Node *node" "M_Real theta"
.Pp
.Ft "void"
.Fn SG_RotateJ "SG_Node *node" "M_Real theta"
.Pp
.Ft "void"
.Fn SG_RotateK "SG_Node *node" "M_Real theta"
.Pp
.Ft "void"
.Fn SG_Rotatevd "SG_Node *node" "M_Real degrees" "M_Vector3 axis"
.Pp
.Ft "void"
.Fn SG_RotateId "SG_Node *node" "M_Real degrees"
.Pp
.Ft "void"
.Fn SG_RotateJd "SG_Node *node" "M_Real degrees"
.Pp
.Ft "void"
.Fn SG_RotateKd "SG_Node *node" "M_Real degrees"
.Pp
.Ft "void"
.Fn SG_GetNodeTransform "void *node" "M_Matrix44 *T"
.Pp
.Ft "void"
.Fn SG_GetNodeTransformInverse "void *node" "M_Matrix44 *T"
.Pp
.Ft "M_Vector3"
.Fn SG_NodePos "SG_Node *node"
.Pp
.Ft "M_Vector3"
.Fn SG_NodeDir "SG_Node *node"
.Pp
.Ft "M_Real"
.Fn SG_NodeSize "SG_Node *node"
.Pp
.nr nS 1
.Fn SG_Identity
sets the transformation matrix of the node to the identity matrix.
.Pp
The
.Fn SG_Translate*
functions multiply
.Va T
by a translation matrix.
.Pp
The
.Fn SG_Scale
function multiplies
.Va T
by a uniform scaling matrix.
.Pp
.Fn SG_Rotate*
multiply
.Va T
by a rotation matrix.
Angles are given in radians, except for
.Fn SG_Rotate*d
variants which accept angular arguments in degrees.
.Pp
.Fn SG_Rotatev
generates a rotation of
.Fa theta
radians around
.Fa axis .
.Pp
Note that most of the preceding functions are trivial wrappers around
.Xr M_Matrix 3
functions (applied to the transformation matrix
.Va T
of the node).
.Pp
The
.Fn SG_GetNodeTransform
function returns a transformation matrix mapping the node back to world
coordinates (i.e., by computing the product of the transformation matrices
of the node and its parents).
.Fn SG_GetNodeTransformInverse
returns the inverse of this matrix.
.Pp
The
.Fn SG_NodePos
function returns a vector representing the absolute world coordinates of
a node.
.Fn SG_NodeDir
returns a normalized vector representing the direction of a node with respect
to the world Z axis (i.e., the Z axis of the origin node).
.Fn SG_NodeSize
returns the absolute scaling factor of an object.
.Sh STRUCTURE DATA
For the
.Nm
object:
.Pp
.Bl -tag -compact -width "M_Matrix44 T "
.It Ft Uint flags
Option flags (see
.Dq FLAGS
section).
.It Ft SG *sg
Back pointer to parent
.Xr SG 3
object.
.It Ft M_Matrix44 T
Transformation matrix (relative to parent node).
.El
.Sh FLAGS
For the
.Nm
object:
.Bl -tag -width "SG_NODE_SELECTED "
.It SG_NODE_SELECTED
Node is selected (e.g., for edition).
.El
.Sh EXAMPLES
See
.Pa sg/sg_dummy.c
in the Agar source distribution for an example node class implementation.
.Sh SEE ALSO
.Xr M_Matrix 3 ,
.Xr M_Real 3 ,
.Xr M_Vector 3 ,
.Xr SG 3 ,
.Xr SG_Intro 3
.Sh HISTORY
The
.Nm
node class first appeared in Agar 1.6.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- The Agar Interface Builder [**agarib**](https://libagar.org/man1/agarib).
- The Agar Debugger [**agardb**](https://libagar.org/man1/agardb).
- Ada: Add Widget classes to the new Ada bindings in `ada/gui/`.
- New COBOL bindings: [GnuCOBOL](https://sourceforge.net/projects/open-cobol/) and [GuiCOBOL](http://www.opencobol.altervista.org).
- [**AG_Console**](https://libagar.org/man3/AG_Console): Handle ansi cursor and terminal operations. Clipboard integration.
- [**AG_FontSelector**](https://libagar.org/man3/AG_FontSelector): Display only those styles that are available for a given font (as opposed to a static list of styles).
- [**AG_Notebook**](https://libagar.org/man3/AG_Notebook): Fix padding issues. Add disposition modes Bottom, Left & Right.
- [**AG_Pixmap**](https://libagar.org/man3/AG_Pixmap) & [**AG_Fixed**](https://libagar.org/man3/AG_Fixed): Zoom operations.
- [**AG_StyleEditor**](https://libagar.org/man3/AG_StyleEditor): Edition of stylesheet rules.
- [**AG_StyleSheet**](https://libagar.org/man3/AG_StyleSheet): Gradients and border styles. New selector "Class > #id" for specific named instances of attached child widgets.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Filters. Stencil operations.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Provide a way to map core font names to known system fonts so porters have the option of adding dependencies to existing font packages in order to reduce the size of agar packages.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) & [**AG_Editable**](https://libagar.org/man3/AG_Editable): Extend SGR support. Syntax highlighting & rich-text editing methods.
- [**AG_WidgetPrimitives**](https://libagar.org/man3/AG_WidgetPrimitives): Dithering. Shadow effects.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Provide a variation of the "zoom" feature to resize widgets individually.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Threading optimizations. Cache the `VISIBLE` flag into an `int` so that culling / rendering and mouse-intersection tests can skip over invisible widgets without locking them.
- [**AG_Window**](https://libagar.org/man3/AG_Window): MRU API to simplify the process of remembering geometries. New gravity methods for autoplacing.
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Cursor operations. Clipboard integration.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Make drag-and-drop (`sockets` test) work correctly (possibly using `AG_WindowMove()`).
- [**wgl**](https://libagar.org/man3/AG_DriverWGL): Clipboard integration. Live resize.
- Fix [**Style Editor**](https://libagar.org/man3/AG_StyleEditor) and [**GUI Debugger**](https://libagar.org/man3/AG_GuiDebugger) crashes which would occur when attempting to iterate over widgets (in keyrepeat mode).
- Build system: Add [**CMake**](https://cmake.org) support.
- A software framebuffer driver for X Windows.

## [1.7.0] - 2022-10-
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**Single-window drivers**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required. Improved appearance of `AG_MenuView`.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): When an `AG_MenuView` is detached, immediately invalidate the `view` pointers of its `AG_MenuItem`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Text**](https://libagar.org/man3/AG_Text): The surface returned by `AG_TextRender()` now includes Ascent information under `Guide 0`.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) and [**AG_Editable**](https://libagar.org/man3/AG_Editable): Implement Undo/Redo functionality.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Draw lines to express the relationship between tree items.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_EXPAND_NODES`. Display nodes (items with children) initially in expanded state.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_NO_SCALE_ICON`. Disable icon scaling when the icon width exceeds the standard width.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_FIXED_HEIGHT`. Don't reset the item height on "font-changed".
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New item flag `AG_TLIST_ITEM_DISABLED`. Disable the item individually (prevent selection and display in "disabled" state).
- [**AG_Event**](https://libagar.org/man3/AG_Event): New macro [AG_EVENT_DUMP](https://libagar.org/man3/AG_EVENT_DUMP) to produce a listing of `argv[]` arguments on the console.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New flag `AG_WIDGET_DISABLE_ON_ATTACH` and functions `AG_PushDisabledState()` and `AG_PopDisabledState`. Make attached child widgets start in DISABLED mode (without raising "widget-disabled").

sg/sg_node.c
/*
 * Copyright (c) 2005-2019 Julien Nadeau Carriere <vedge@csoft.net>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Base class for any node in a SG(3) scene graph.
 */

#include <agar/core/core.h>
#include <agar/sg/sg.h>
#include <agar/sg/sg_gui.h>

#include <string.h>

static void
OnAttach(AG_Event *_Nonnull event)
{
	SG_Node *node = SG_NODE_SELF();
	AG_Object *parent = AG_OBJECT_PTR(1);
	SG *sg = NULL;
	
	if (AG_OfClass(parent, "SG:*")) {
		sg = (SG *)parent;
	} else if (AG_OfClass(parent, "SG_Node:*")) {
		sg = SGNODE(parent)->sg;
	} else {
		AG_FatalError("Invalid SG_Node parent");
	}
	AG_ObjectLock(sg);
	TAILQ_INSERT_TAIL(&sg->nodes, node, nodes);
	node->sg = sg;
	AG_ObjectUnlock(sg);
}

static void
OnDetach(AG_Event *_Nonnull event)
{
	SG_Node *node = SG_NODE_SELF();
	void *parent = AG_OBJECT_PTR(1);
	SG *sg = NULL;
	
	if (AG_OfClass(parent, "SG:*")) {
		sg = (SG *)parent;
	} else if (AG_OfClass(parent, "SG_Node:*")) {
		sg = SGNODE(parent)->sg;
	} else {
		AG_FatalError("Invalid SG_Node parent");
	}
	AG_ObjectLock(sg);
	TAILQ_REMOVE(&sg->nodes, node, nodes);
	node->sg = NULL;
	AG_ObjectUnlock(sg);
}

static void
Init(void *_Nonnull obj)
{
	SG_Node *node = obj;

	node->flags = 0;
	M_MatIdentity44v(&node->T);
	TAILQ_INIT(&node->actions);

	AG_SetEvent(node, "attached", OnAttach, NULL);
	AG_SetEvent(node, "detached", OnDetach, NULL);
}

static void
Reset(void *_Nonnull obj)
{
	SG_Node *node = obj;
	SG_Action *act, *actNext;

	for (act = TAILQ_FIRST(&node->actions);
	     act != TAILQ_END(&node->actions);
	     act = actNext) {
		actNext = TAILQ_NEXT(act, actions);
		Free(act);
	}
	TAILQ_INIT(&node->actions);
}

static int
Load(void *_Nonnull obj, AG_DataSource *_Nonnull buf, const AG_Version *_Nonnull ver)
{
	SG_Node *node = obj;

	node->flags = (Uint)AG_ReadUint32(buf);
	M_ReadMatrix44v(buf, &node->T);
	return (0);
}

static int
Save(void *_Nonnull obj, AG_DataSource *_Nonnull buf)
{
	SG_Node *node = obj;

	AG_WriteUint32(buf, (Uint32)node->flags);
	M_WriteMatrix44(buf, &node->T);
	return (0);
}

static void *_Nullable
Edit(void *_Nonnull obj, SG_View *_Nullable sgv)
{
	SG_Node *node = obj;
	AG_Box *box;

	box = AG_BoxNewVert(NULL, AG_BOX_HFILL);
	AG_LabelNew(box, 0, _("Node Name: %s"), AGOBJECT(node)->name);
	return (box);
}

/* Generic edit() operation for node objects. */
void *_Nonnull
SG_NodeEdit(void *obj)
{
	SG_Node *node = obj;
	AG_Window *win;
	void *wEdit;
	AG_ObjectClass **hier;
	int i, nHier;

	if ((win = AG_WindowNew(0)) == NULL) {
		AG_FatalError(NULL);
	}
	AG_WindowSetCaptionS(win, OBJECT(node)->name);
	
	if (AG_ObjectGetInheritHier(node, &hier, &nHier) != 0) {
		AG_FatalError(NULL);
	}
	for (i = 0; i < nHier; i++) {
		SG_NodeClass *nc = (SG_NodeClass *)hier[i];

		if (!AG_ClassIsNamed(hier[i], "SG_Node:*") ||
		    nc->edit == NULL)
			continue;

		wEdit = nc->edit(node, NULL);
		if (wEdit != NULL && AG_OfClass(wEdit, "AG_Widget:*"))
			AG_ObjectAttach(win, wEdit);
	}
	Free(hier);
	return (win);
}

/*
 * Compute the product of the transform matrices of the given node
 * and its parents.
 */
void
SG_GetNodeTransform(void *obj, M_Matrix44 *T)
{
	SG_Node *node = obj;
#ifdef AG_THREADS
	SG *sg = node->sg;
#endif
	SG_Node *chld = node;
	TAILQ_HEAD_(sg_node) rnodes = TAILQ_HEAD_INITIALIZER(rnodes);

	M_MatIdentity44v(T);

	AG_ObjectLock(sg);
	while (chld != NULL) {
		TAILQ_INSERT_TAIL(&rnodes, chld, rnodes);
		if (OBJECT(chld)->parent == NULL ||
		    !AG_OfClass(OBJECT(chld)->parent, "SG_Node:*")) {
			break;
		}
		chld = OBJECT(chld)->parent;
	}
	TAILQ_FOREACH(chld, &rnodes, rnodes) {
		M_MatMult44v(T, &chld->T);
	}
	AG_ObjectUnlock(sg);
}

/*
 * Compute the product of the inverse transform matrices of the given node
 * and its parents.
 */
void
SG_GetNodeTransformInverse(void *obj, M_Matrix44 *T)
{
	SG_Node *node = obj;
	SG_Node *chld = node;
#ifdef AG_THREADS
	SG *sg = node->sg;
#endif
	TAILQ_HEAD_(sg_node) rnodes = TAILQ_HEAD_INITIALIZER(rnodes);

	M_MatIdentity44v(T);

	AG_ObjectLock(sg);
	while (chld != NULL) {
		TAILQ_INSERT_TAIL(&rnodes, chld, rnodes);
		if (OBJECT(chld)->parent == NULL ||
		    !AG_OfClass(OBJECT(chld)->parent, "SG_Node:*")) {
			break;
		}
		chld = OBJECT(chld)->parent;
	}
	TAILQ_FOREACH(chld, &rnodes, rnodes) {
		M_Matrix44 Tinv;

		Tinv = M_MatInvert44(chld->T);
		M_MatMult44v(T, &Tinv);
	}
	AG_ObjectUnlock(sg);
}

/*
 * Graphically render a node and its children.
 * The SG must be locked.
 */
void
SG_NodeDraw(SG *sg, SG_Node *node, SG_View *view)
{
	AG_ObjectClass **hier;
	int i, nHier;
	M_Matrix44 Tsave, T;
	SG_Node *chld;

	GL_FetchMatrixv(GL_MODELVIEW_MATRIX, &Tsave);
	T = M_MatTranspose44p(&node->T);	/* OpenGL is column-major */
	GL_MultMatrixv(&T);

	AG_ObjectLock(node);

	/* Render this node. */
	if (AG_ObjectGetInheritHier(node, &hier, &nHier) != 0) {
		AG_FatalError(NULL);
	}
	for (i = nHier-1; i >= 0; i--) {
		SG_NodeClass *nc = (SG_NodeClass *)hier[i];

		if (!AG_ClassIsNamed(hier[i], "SG_Node:*") ||
		    nc->draw == NULL) {
			continue;
		}
		nc->draw(node, view);
		break;
	}

	/* Render child nodes. */
	OBJECT_FOREACH_CLASS(chld, node, sg_node, "SG_Node:*") {
		SG_NodeDraw(sg, chld, view);
	}
	AG_ObjectUnlock(node);

	GL_LoadMatrixv(&Tsave);
	Free(hier);
}

/* Save node data (and child nodes) to a data source. */
int
SG_NodeSave(SG *sg, AG_DataSource *ds, SG_Node *node)
{
	off_t countOffs, skipSizeOffs;
	Uint32 count;
	SG_Node *chldNode;

	AG_ObjectLock(node);

	/* Save object metadata */
	AG_WriteString(ds, OBJECT(node)->name);
	if (OBJECT_CLASS(node)->pvt.libs[0] != '\0') {
		char s[AG_OBJECT_TYPE_MAX];
		Strlcpy(s, OBJECT_CLASS(node)->hier, sizeof(s));
		Strlcat(s, "@", sizeof(s));
		Strlcat(s, OBJECT_CLASS(node)->pvt.libs, sizeof(s));
		AG_WriteString(ds, s);
	} else {
		AG_WriteString(ds, OBJECT_CLASS(node)->hier);
	}
	skipSizeOffs = AG_Tell(ds);
	AG_WriteUint32(ds, 0);

	/* Save object dataset */
	if (AG_ObjectSerialize(node, ds) == -1)
		goto fail;

	/* Save child nodes */
	countOffs = AG_Tell(ds);
	AG_WriteUint32(ds, 0);
	count = 0;
	OBJECT_FOREACH_CHILD(chldNode, node, sg_node) {
		if (SG_NodeSave(sg, ds, chldNode) == -1) {
			goto fail;
		}
		count++;
	}

	AG_WriteUint32At(ds, count, countOffs);
	AG_WriteUint32At(ds, AG_Tell(ds)-skipSizeOffs, skipSizeOffs);

	AG_ObjectUnlock(node);
	return (0);
fail:
	AG_ObjectUnlock(node);
	return (-1);
}

/* Load node data (and children) from data source. */
int
SG_NodeLoad(SG *sg, AG_DataSource *ds, SG_Node *parentNode)
{
	Uint count, i;
	char nodeName[AG_OBJECT_NAME_MAX];
	char classSpec[AG_OBJECT_TYPE_MAX];
	AG_ObjectClass *nodeClass;
	/* Uint32 skipSize; */
	SG_Node *node;

	/* Load object metadata */
	AG_CopyString(nodeName, ds, sizeof(nodeName));
	AG_CopyString(classSpec, ds, sizeof(classSpec));
/*	skipSize = AG_ReadUint32(ds); */
	(void)AG_ReadUint32(ds);
	
	/* Allocate node instance; load DSOs as needed. */
	if ((nodeClass = AG_LoadClass(classSpec)) == NULL) {
		return (-1);
	}
	if ((node = AG_TryMalloc(nodeClass->size)) == NULL) {
		return (-1);
	}
	AG_ObjectInit(node, nodeClass);
	AG_ObjectSetNameS(node, nodeName);

	/* Load object dataset */
	if (AG_ObjectUnserialize(node, ds) == -1)
		goto fail;

	/* Attach node to parent. */
	if (parentNode != NULL) {
		AG_ObjectAttach(parentNode, node);
	} else {
		AG_ObjectAttach(sg, node);
		sg->root = node;
	}

	/* Load child nodes */
	count = (Uint)AG_ReadUint32(ds);
	for (i = 0; i < count; i++) {
		if (SG_NodeLoad(sg, ds, node) == -1)
			goto fail_detach;
	}
	return (0);
fail_detach:
	if (node == sg->root) { sg->root = NULL; }
fail:
	AG_ObjectDestroy(node);
	return (-1);
}

SG_NodeClass sgNodeClass = {
	{
		"SG_Node",
		sizeof(SG_Node),
		{ 0,0 },
		Init,
		Reset,
		NULL,		/* destroy */
		Load,
		Save,
		SG_NodeEdit
	},
	NULL,			/* menuInstance */
	NULL,			/* menuClass */
	NULL,			/* draw */
	NULL,			/* intersect */
	Edit
};
